### PR TITLE
Add reports for edtf and w3cdtf validation

### LINKED ISSUE #4071: Report on invalid EDTF dates in Cocina description
This is to check remediation progress for invalid EDTF dates. For valid date patterns, see #3797.

Desired output: druid, collection druid, catkey, date value(s).

Check for values in description..date..value with description..date.encoding.code edtf.

### LINKED ISSUE #4073: Report on invalid W3CDTF dates in Cocina description
This is to check remediation progress for invalid W3CDTF dates. For valid date patterns see #3848.

Desired output: druid, collection druid, catkey, date value(s).

Check for values in description..date..value with description..date.encoding.code w3cdtf.

### PR DESCRIPTION
**Pull request by mjgiarlo**
## Why was this change made? 🤔

Connects to #4071
Connects to #4073

Also brings the ISO8601 date report into alignment with the new reports and dumps bespoke date validation in favor of using upstream libraries.

## How was this change tested? 🤨

Ran all three reports on stage

### CODE REVIEW

File: app/reports/bad_iso8601_dates.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  def collection_id
+    dro.structural['isMemberOf'].first
+  end
+
+  def catkey
+    dro.identification['catalogLinks'].find { |link| link['catalog'] == 'symphony' }&.fetch('catalogRecordId')
+  end
+end
```
Comment (mjgiarlo) [inline app/reports/bad_iso8601_dates.rb:46]: I refactored and renamed this report, @jcoyne, to align with the other two date reports and to accommodate a couple changes:

1. Report on all bad date values, not just the first
1. Report on values that are nested in structured and parallel and grouped values

**Review by ndushay — COMMENTED** [review]
So this addresses #4071, #4072, #4072 and #4052?

How does it cope with Arcadia's desire for what may be a subset of the allowed regex values in the etdf gem, for example?

How will this relate to the date validation code going into cocina when someone picks up sul-dlss/cocina-models#483?   I see you remove that code here.  Also, we're going to do date validation via cocina-models so it seems prudent to have these reports run through the same validation code (I know,  sul-dlss/cocina-models#483 didn't exist when you picked up this ticket ...)

Beyond those questions, the code here looks fine to me.  Confer with JCoyne and Arcadia, I guess.

**Comment by mjgiarlo** [comment]
@ndushay 
> So this addresses #4071, #4072, #4072 and #4052?

And #4073, yeah. I wasn't aware of #4052 when I started this PR, and I can remove those bits if folks prefer; I don't need it for these reports. I'm flexible.
 
> How does it cope with Arcadia's desire for what may be a subset of the allowed regex values in the etdf gem, for example?

This is the first I've heard of it. Can you or @arcadiafalcone point me at more information?
 
> How will this relate to the date validation code going into cocina when someone picks up [sul-dlss/cocina-models#483](https://github.com/sul-dlss/cocina-models/issues/483)? I see you remove that code here ...

Once it's in Cocina and that's deployed, entering bad dates will no longer be possible, but we may still have data to remediate. I'd say once all three reports run clean, we can ditch these reports and the changes in this commit.
 
> Beyond those questions, the code here looks fine to me. Confer with JCoyne and Arcadia, I guess.

cc: @jcoyne @arcadiafalcone 😄

File: app/reports/bad_iso8601_dates.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  def initialize(dro:)
+    @dro = dro
+  end
+
+  def report
+    bad_values = path.on(dro.description.to_json).filter_map do |date_value|
+      DateTime.iso8601(date_value)
+      nil
+    rescue ArgumentError
```
Comment (jcoyne) [inline app/reports/bad_iso8601_dates.rb:22]: Shouldn't this be a `Date::Error`?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/reports/bad_iso8601_dates.rb (hunk 2 above)
Comment (mjgiarlo) [reply app/reports/bad_iso8601_dates.rb:22]: Yes, copypasta. Thank you!

**Review by jcoyne — APPROVED** [review]

**Comment by ndushay** [comment]
@mjgiarlo can you easily determine which date patterns are allowed by this code?   The regexp in the date validator class (which will be migrating to cocina models)  should reflect what @arcadiafalcone wanted.  I don't know if it's a subset or not ... it's just that we've got more of a black box on valid dates with this implementation.

**Comment by jcoyne** [comment]
@ndushay a Regex is not sufficient for validating dates (unless it's monstrously complex).  Consider leap years, Sept 31, etc.

**Comment by mjgiarlo** [comment]
@ndushay @jcoyne I chatted with Arcadia and she is on board with using upstream validators ([comment](https://github.com/sul-dlss/cocina-models/issues/392#issuecomment-1150440283)). @ndushay can you re-review?

**Review by ndushay — APPROVED** [review]

**Pull request merged**